src/serac/numerics/functional/detail/hexahedron_Hcurl.inl

```diff
@@ -166,5 +248,612 @@ struct finite_element<Geometry::Hexahedron, Hcurl<p>> {
 
     return curl;
   }
+
+  template <typename in_t, int q>
+  static auto batch_apply_shape_fn(int j, tensor<in_t, q * q * q> input, const TensorProductQuadratureRule<q>&)
+  {
+    constexpr bool                     apply_weights = false;
+    constexpr tensor<double, q, p>     B1            = calculate_B1<apply_weights, q>();
+    constexpr tensor<double, q, p + 1> B2            = calculate_B2<apply_weights, q>();
+    constexpr tensor<double, q, p + 1> G2            = calculate_G2<apply_weights, q>();
+
+    // figure out which node and which direction
+    // correspond to the dof index "j"
+    int jx, jy, jz;
+    int dir       = j / (p * (p + 1) * (p + 1));
+    int remainder = j % (p * (p + 1) * (p + 1));
+    switch (dir) {
+      case 0:  // x-direction
+        jx = remainder % p;
+        jy = (remainder % (p * (p + 1))) / p;
+        jz = remainder / (p * (p + 1));
+        break;
+
+      case 1:  // y-direction
+        jx = remainder % (p + 1);
+        jy = (remainder % (p * (p + 1))) / (p + 1);
+        jz = remainder / (p * (p + 1));
+        break;
+
+      case 2:  // z-direction
+        jx = remainder % (p + 1);
+        jy = (remainder % ((p + 1) * (p + 1))) / (p + 1);
+        jz = remainder / ((p + 1) * (p + 1));
+        break;
+    }
+
+    using vec3     = tensor<double, 3>;
+    using source_t = decltype(dot(get<0>(get<0>(in_t{})), vec3{}) + dot(get<1>(get<0>(in_t{})), vec3{}));
+    using flux_t   = decltype(dot(get<0>(get<1>(in_t{})), vec3{}) + dot(get<1>(get<1>(in_t{})), vec3{}));
+
+    tensor<tuple<source_t, flux_t>, q * q * q> output;
+
+    for (int qz = 0; qz < q; qz++) {
+      for (int qy = 0; qy < q; qy++) {
+        for (int qx = 0; qx < q; qx++) {
+          tensor<double, 3> phi_j{};
+          tensor<double, 3> curl_phi_j{};
+
+          switch (dir) {

```

Comment:
Yes. This is reminding me why I hate Hcurl!

Reply:
There's a different dof numbering that cleans this up considerably, but that's not the one mfem uses.